examples/src/state-migration.ts

```diff
@@ -0,0 +1,70 @@
+import { NearBindgen, view, near, migrate, call, Vector, assert } from 'near-sdk-js'
+import { AccountId } from 'near-sdk-js/lib/types'
+
+type OldMessageFormat = {
+    sender: AccountId,
+    header: string,
+    text: string,
+}
+
+type NewMessageFormat = {
+    sender: AccountId,
+    recipient?: AccountId,
+    header: string,
+    text: string,
+}
+
+@NearBindgen({})
+export class MigrationDemo {
+    messages: Vector<NewMessageFormat>;
+
+    constructor() {
+        this.messages = new Vector<NewMessageFormat>('messages');
+    }
+
+    @call({ payableFunction: true })
+    addMessage({ message }: {
+        message: NewMessageFormat
+    }): void {
+        this.messages.push(message);
+        near.log(`${near.signerAccountId()} added message ${JSON.stringify(message)}`);
+    }
+
+    @view({})
+    countMessages(): number {
+        return this.messages.toArray().length;
+    }
+
+
+    @migrate({})
+    migrateState(): Vector<NewMessageFormat> {
+        assert(this.messages.toArray().length == 0, "Contract state should not be deserialized in @migrate");
+        // retrieve the current state from the contract
+        let raw_vector = JSON.parse(near.storageRead("STATE")).messages;
```

Comment:
Instead of reading from raw state, `migrateState` can takes an argument and `@migrate({})` does deserialization for it. e.g.:

```
    migrateState({messages}: {messages: Vector<OldMessageFormat>}): Vector<NewMessageFormat> {
        let old_messages = messages;
        // same as your implementation
```
```

Reply:
Why should we pass old messages?
I think that the whole idea behind migration is to rewrite the data in new format. It means:
- read the old data
- reformat it
- save new data to state

Reply:
Your approach works and looks intuitive! I was thinking a syntax sugar to read the old data to avoid `JSON.parse(near.storageRead("STATE"))` pattern. User can write:
```
@Migrate
migrateState(oldState: OldStateFormat): NewStateFormat {
    let newState: NewStateFormat = some_reformat_steps(oldState);
    return newState;
}
```
`@Migrate` would know how to load (deserialize) oldState given OldStateFormat. `@Migrate` would also serialize reformat state using newStateFormat.

A possible `@Migrates`'s deserialize:
```
let oldRaw = near.storageRead("STATE")
let old = OldStateFormat.deserialize ? OldStateFormat.deserialize(oldRaw) 
  : OldStateFormat.reconstruct ? OldStateFormat.reconstruct(JSON.parse(oldRaw))
  : JSON.parse(oldRaw)
```

Reply:
^ is i think nice to have in API but it's up to your preference. The PR is considered done!

Reply:
Since we do not have any NEP - or complex approaches are questionable. Let's get back to it if necessary.